src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,337 @@
+            var removeUnreadMembersRule = CreateDescriptor(
+                IDEDiagnosticIds.RemoveUnreadMembersDiagnosticId, removeUnreadMembersTitle, removeUnreadMembersMessage, configurable: true, enabledByDefault: forceEnableRules);
+            var removeUnreadMembersRuleUnnecessaryWithFadingRule = CreateUnnecessaryDescriptor(
+                IDEDiagnosticIds.RemoveUnreadMembersDiagnosticId, removeUnreadMembersTitle, removeUnreadMembersMessage, configurable: true, enabledByDefault: forceEnableRules);
+
+            return ImmutableArray.Create(removeUnusedMembersRule, removeUnusedMembersRuleWithFadingRule,
+                    removeUnreadMembersRule, removeUnreadMembersRuleUnnecessaryWithFadingRule);
+        }
+
+        private DiagnosticDescriptor RemoveUnusedMemberRule => SupportedDiagnostics[1];
+        private DiagnosticDescriptor RemoveUnreadMemberRule => SupportedDiagnostics[3];
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        // We need to analyze the whole document even for edits within a method body.
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            // We want to analyze references in generated code, but not report unused members in generated code.
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+
+            context.RegisterCompilationStartAction(compilationStartContext =>
+            {
+                var compilationAnalyzer = new CompilationAnalyzer(compilationStartContext.Compilation, RemoveUnusedMemberRule, RemoveUnreadMemberRule);
+
+                // We register following actions in the compilation:
+                // 1. A symbol action for member symbols to ensure the member's unused state is initialized to true for every private member symbol.
+                // 2. Operation actions for member references and invocations to detect member usages, i.e. read or read reference taken.
+                // 3. Operation action for field initializers to detect non-constant initialization.
+                // 4. Operation action for invalid operations to bail out on erroneous code.
+                // 5. A symbol start/end action for named types to report diagnostics for candidate members that have no usage in executable code.
+                //
+                // Note that we need to register separately for OperationKind.Invocation due to https://github.com/dotnet/roslyn/issues/26206
+
+                compilationStartContext.RegisterSymbolAction(compilationAnalyzer.AnalyzeSymbolDeclaration, SymbolKind.Method, SymbolKind.Field, SymbolKind.Property, SymbolKind.Event);
```

Comment:
you could be even cuter and ask the CompilationAnalyzer to register.  Then, all these members would become private, and only teh CompilationAnalyzer woudl control how it hooked up to the engine.

Reply:
Good idea! Done.

---
In reply to: [213413533](https://github.com/dotnet/roslyn/pull/29511#discussion_r213413533) [](ancestors = 213413533)